File: ruleforge/parser/parser.py
from ..lexer import TokenType
from .ast_nodes import RuleNode, ActionNode, BinaryOpNode, NullCheckNode, LiteralNode, PropertyAccessNode, FunctionCallNode
from .errors import ParserError

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0
        self.current_token = self.tokens[self.pos]

    def advance(self):
        self.pos += 1
        if self.pos < len(self.tokens):
            self.current_token = self.tokens[self.pos]

    def expect(self, token_type):
        if self.current_token.type == token_type:
            self.advance()
        else:
            t = self.current_token
            raise ParserError("RF2002", f"Expected {token_type} but got {t.type} ('{t.value}')", t.line, t.column)

    def parse(self):
        rules = []
        while self.current_token.type != TokenType.EOF:
            rules.append(self.parse_rule())
        return rules

    def parse_rule(self):
        self.expect(TokenType.RULE)
        rule_name = self.current_token.value
        self.expect(TokenType.IDENTIFIER)
        
        self.expect(TokenType.LANGUAGE)
        lang_version = int(self.current_token.value)
        self.expect(TokenType.INTEGER)
        
        self.expect(TokenType.WHEN)
        when_expr = self.expression()
        
        self.expect(TokenType.THEN)
        then_actions = self.action_list()
        
        else_actions = []
        if self.current_token.type == TokenType.ELSE:
            self.advance()
            else_actions = self.action_list()
            
        self.expect(TokenType.END)
        return RuleNode(rule_name, lang_version, when_expr, then_actions, else_actions)

    def action_list(self):
        actions = []
        actions.append(self.parse_action())
        while self.current_token.type in [TokenType.ALLOW, TokenType.DENY, TokenType.ALERT, TokenType.APPLY, TokenType.NO_ACTION]:
            actions.append(self.parse_action())
            
        # RF2005 Invalid Action (NO_ACTION must be alone)
        if TokenType.NO_ACTION in [a.action_type for a in actions] and len(actions) > 1:
            t = self.current_token
            raise ParserError("RF2005", "NO_ACTION must be the only action in its block", t.line, t.column)
            
        return actions

    def parse_action(self):
        token = self.current_token
        if token.type in [TokenType.ALLOW, TokenType.NO_ACTION]:
            self.advance()
            return ActionNode(token.value)
        elif token.type in [TokenType.DENY, TokenType.ALERT, TokenType.APPLY]:
            self.advance()
            val_token = self.current_token
            self.expect(TokenType.STRING)
            return ActionNode(token.value, val_token.value)
        raise ParserError("RF2005", f"Expected action but got {token.type}", token.line, token.column)

    # Precedence: OR -> AND -> NOT -> Comparison -> Add/Sub -> Mul/Div -> Primary
    def expression(self):
        return self.logical_or()

    def logical_or(self):
        node = self.logical_and()
        while self.current_token.type == TokenType.OR:
            op = self.current_token.value; self.advance()
            node = BinaryOpNode(node, op, self.logical_and())
        return node

    def logical_and(self):
        node = self.logical_not()
        while self.current_token.type == TokenType.AND:
            op = self.current_token.value; self.advance()
            node = BinaryOpNode(node, op, self.logical_not())
        return node

    def logical_not(self):
        if self.current_token.type == TokenType.NOT:
            self.advance()
            return BinaryOpNode(None, "NOT", self.logical_not())
        return self.comparison()

    def comparison(self):
        node = self.arithmetic()
        
        if self.current_token.type == TokenType.IS:
            self.advance()
            is_not = False
            if self.current_token.type == TokenType.NOT:
                self.advance()
                is_not = True
            self.expect(TokenType.NULL)
            return NullCheckNode(node, is_not)
            
        while self.current_token.type in [TokenType.EQ, TokenType.NEQ, TokenType.GT, TokenType.LT, TokenType.GTE, TokenType.LTE]:
            op = self.current_token.value; self.advance()
            node = BinaryOpNode(node, op, self.arithmetic())
        return node

    def arithmetic(self):
        node = self.term()
        while self.current_token.type in [TokenType.PLUS, TokenType.MINUS]:
            op = self.current_token.value; self.advance()
            node = BinaryOpNode(node, op, self.term())
        return node

    def term(self):
        node = self.factor()
        while self.current_token.type in [TokenType.MULTIPLY, TokenType.DIVIDE]:
            op = self.current_token.value; self.advance()
            node = BinaryOpNode(node, op, self.factor())
        return node

    def factor(self):
        token = self.current_token
        
        if token.type == TokenType.LPAREN:
            self.advance()
            expr = self.expression()
            self.expect(TokenType.RPAREN)
            return expr
            
        elif token.type in [TokenType.INTEGER, TokenType.DECIMAL, TokenType.STRING, TokenType.BOOLEAN, TokenType.DATE]:
            self.advance()
            return LiteralNode(token.value, token.type)
            
        elif token.type == TokenType.IDENTIFIER:
            name = token.value
            self.advance()
            
            if self.current_token.type == TokenType.LPAREN:
                self.advance()
                args = []
                if self.current_token.type != TokenType.RPAREN:
                    args.append(self.expression())
                    while self.current_token.type == TokenType.COMMA:
                        self.advance()
                        args.append(self.expression())
                self.expect(TokenType.RPAREN)
                return FunctionCallNode(name, args)
                
            elif self.current_token.type == TokenType.DOT:
                self.advance()
                prop_token = self.current_token
                self.expect(TokenType.IDENTIFIER)
                return PropertyAccessNode(name, prop_token.value)
                
            return LiteralNode(token.value, token.type) 
            
        elif token.type == TokenType.EOF:
            raise ParserError("RF2003", "Unexpected EOF, expected expression", token.line, token.column)
            
        raise ParserError("RF2004", f"Invalid expression, unexpected token {token.type}", token.line, token.column)

File: dotnet/src/RuleForge.Core/Parser/Parser.cs
using RuleForge.Core.Lexing;
using RuleForge.Core.Syntax;

namespace RuleForge.Core.Parser;

public class Parser
{
    private readonly List<Token> _tokens;
    private int _pos;
    private Token CurrentToken => _tokens[_pos];

    public Parser(List<Token> tokens)
    {
        _tokens = tokens;
        _pos = 0;
    }

    private void Advance() => _pos++;
    
    private Token Expect(TokenType type)
    {
        if (CurrentToken.Type == type)
        {
            var t = CurrentToken;
            Advance();
            return t;
        }
        throw new ParserException("RF2002", $"Expected {type} but got {CurrentToken.Type} ('{CurrentToken.Value}')", CurrentToken.Line, CurrentToken.Column);
    }

    public List<RuleNode> Parse()
    {
        var rules = new List<RuleNode>();
        while (CurrentToken.Type != TokenType.EOF)
        {
            rules.Add(ParseRule());
        }
        return rules;
    }

    private RuleNode ParseRule()
    {
        Expect(TokenType.RULE);
        var name = CurrentToken.Value;
        Expect(TokenType.IDENTIFIER);
        
        Expect(TokenType.LANGUAGE);
        if (!int.TryParse(CurrentToken.Value, out int langVer))
            throw new ParserException("RF2002", "Expected integer for language version", CurrentToken.Line, CurrentToken.Column);
        Expect(TokenType.INTEGER);
        
        Expect(TokenType.WHEN);
        var whenExpr = Expression();
        
        Expect(TokenType.THEN);
        var thenActions = ActionList();
        
        var elseActions = new List<ActionNode>();
        if (CurrentToken.Type == TokenType.ELSE)
        {
            Advance();
            elseActions = ActionList();
        }
        
        Expect(TokenType.END);
        return new RuleNode(name, langVer, whenExpr, thenActions, elseActions);
    }

    private List<ActionNode> ActionList()
    {
        var actions = new List<ActionNode>();
        actions.Add(ParseAction());
        
        while (CurrentToken.Type == TokenType.ALLOW || CurrentToken.Type == TokenType.DENY || 
               CurrentToken.Type == TokenType.ALERT || CurrentToken.Type == TokenType.APPLY || 
               CurrentToken.Type == TokenType.NO_ACTION)
        {
            actions.Add(ParseAction());
        }
        return actions;
    }

    private ActionNode ParseAction()
    {
        var t = CurrentToken;
        if (t.Type == TokenType.ALLOW || t.Type == TokenType.NO_ACTION)
        {
            Advance();
            return new ActionNode(t.Value);
        }
        if (t.Type == TokenType.DENY || t.Type == TokenType.ALERT || t.Type == TokenType.APPLY)
        {
            Advance();
            var valToken = CurrentToken;
            Expect(TokenType.STRING);
            return new ActionNode(t.Value, valToken.Value);
        }
        throw new ParserException("RF2005", $"Expected action but got {t.Type}", t.Line, t.Column);
    }

    private Expression Expression() => LogicalOr();

    private Expression LogicalOr()
    {
        var node = LogicalAnd();
        while (CurrentToken.Type == TokenType.OR)
        {
            var op = CurrentToken.Value;
            Advance();
            node = new BinaryExpression(node, op, LogicalAnd());
        }
        return node;
    }

    private Expression LogicalAnd()
    {
        var node = LogicalNot();
        while (CurrentToken.Type == TokenType.AND)
        {
            var op = CurrentToken.Value;
            Advance();
            node = new BinaryExpression(node, op, LogicalNot());
        }
        return node;
    }

    private Expression LogicalNot()
    {
        if (CurrentToken.Type == TokenType.NOT)
        {
            Advance();
            return new UnaryExpression("NOT", LogicalNot());
        }
        return Comparison();
    }

    private Expression Comparison()
    {
        var node = Arithmetic();
        
        if (CurrentToken.Type == TokenType.IS)
        {
            Advance();
            bool isNot = false;
            if (CurrentToken.Type == TokenType.NOT)
            {
                Advance();
                isNot = true;
            }
            Expect(TokenType.NULL);
            return new NullCheckExpression(node, isNot);
        }
        
        var compTypes = new[] { TokenType.EQ, TokenType.NEQ, TokenType.GT, TokenType.LT, TokenType.GTE, TokenType.LTE };
        if (compTypes.Contains(CurrentToken.Type))
        {
            var op = CurrentToken.Value;
            Advance();
            return new BinaryExpression(node, op, Arithmetic());
        }
        return node;
    }

    private Expression Arithmetic()
    {
        var node = Term();
        while (CurrentToken.Type == TokenType.PLUS || CurrentToken.Type == TokenType.MINUS)
        {
            var op = CurrentToken.Value;
            Advance();
            node = new BinaryExpression(node, op, Term());
        }
        return node;
    }

    private Expression Term()
    {
        var node = Factor();
        while (CurrentToken.Type == TokenType.MULTIPLY || CurrentToken.Type == TokenType.DIVIDE)
        {
            var op = CurrentToken.Value;
            Advance();
            node = new BinaryExpression(node, op, Factor());
        }
        return node;
    }

    private Expression Factor()
    {
        var t = CurrentToken;
        if (t.Type == TokenType.LPAREN)
        {
            Advance();
            var expr = Expression();
            Expect(TokenType.RPAREN);
            return expr;
        }
        if (t.Type == TokenType.INTEGER || t.Type == TokenType.DECIMAL || 
            t.Type == TokenType.STRING || t.Type == TokenType.BOOLEAN || t.Type == TokenType.DATE)
        {
            Advance();
            return new LiteralExpression(t.Value, t.Type);
        }
        if (t.Type == TokenType.IDENTIFIER)
        {
            var name = t.Value;
            Advance();
            if (CurrentToken.Type == TokenType.DOT)
            {
                Advance();
                var propToken = CurrentToken;
                Expect(TokenType.IDENTIFIER);
                return new PropertyExpression(name, propToken.Value);
            }
            return new LiteralExpression(name, t.Type);
        }
        throw new ParserException("RF2004", $"Invalid expression, unexpected token {t.Type}", t.Line, t.Column);
    }
}
